File: LinkedList/Circular LinkedList/Singly LinkedList/CSL.c
#include <stdio.h>
#include <stdlib.h>

// Node data
typedef struct Linkedlist {
	int data;
	struct Linkedlist *next;
} Node;
Node *head, *temp, *node;
int LENGTH = 0;

// Create node function
Node *createNode(int data) {
	Node *temp = (Node *)malloc(sizeof(Node));
	temp->data = data;
	temp->next = temp;
	return temp;
}

// Insert Node
void insert(int index, int data) {
	if (LENGTH==0 && index==1) {  // If list is empty
		node = createNode(data);  // Create node
		head = node;
		LENGTH++;
		printf("%d entered at in begining.", data);
	}
	else if (index>0 && index<=LENGTH+1) {
		node = createNode(data);  // Create node
		temp = head;

		if (index == 1) {   // If index is head
			node->next = head;
			head = node;
			for (int i=1; i<LENGTH; i++)	// Traverse to last node
				temp = temp->next;
			temp->next = head;
		}
		else {      // Other position in list
			for (int i=1; i<index-1; i++)   // Traverse to index-1 node
				temp = temp->next;
			// Making links
			node->next = temp->next;
			temp->next = node;
		}

		LENGTH++;
		printf("%d endered at index %d.", data, index);
	}
	else {
		printf("Invalid Index.\n");
	}
}

// Delete Node
void delete(int index) {
	if (LENGTH == 0) {
		printf("Empty List.\n");
	}
	else {
		if (index>0 && index<=LENGTH) {

			if (LENGTH==index && index==1) {	// If length of list is one
				node = head;
			}
			else if (index == 1) {					// if index is head
				temp = head;
				for (int i=1; i<LENGTH; i++)	// Traverse to last node
					temp = temp->next;
				node = temp->next;
				temp->next = node->next;
				head = node->next;
			}
			else {									// For other index

				temp = head;
				for (int i=1; i<index-1; i++)   // Traverse to index-1 node
					temp = temp->next;
				// Making links
				node = temp->next;
				temp->next = node->next;

			}

			printf("%d is deleted from list.", node->data);
			free(node);
			LENGTH--;

		}
		else {
			printf("Invalid Index.\n");
		}
	}
}

// Display
void display() {
	if (LENGTH == 0) {
		printf("Empty list.\n");
	}
	else {
		temp = head;
		do {
			printf("%-5d",temp->data);
			temp = temp->next;
		} while (temp != head);
		printf("\n");
	}
}

// DRIVER CODE
int main()
{
	
	int switch_var;		// Switch Variable
	int exit = 0;		// Exit condition
	int t1, t2;			// Temporary variable

	while (!exit) {
		// Menu
		printf("\nEnter \n\t1. Insert \n\t2. Delete \n\t3. DISPLAY \n\t4. EXIT \n-> ");
		scanf("%d", &switch_var);

		switch (switch_var) {
			// INSERT
			case 1:
				printf("Enter a value to insert : ");
				scanf("%d", &t1);
				printf("Enter  index(1-%d) : ", LENGTH+1);
				scanf("%d", &t2);
				insert(t2, t1);
				break;
			// DELETE
			case 2:
				printf("Enter index(1-%d) : ", LENGTH);
				scanf("%d", &t1);
				delete(t1);
				break;
			// DISPLAY
			case 3:
				display();
				break;
			// EXIT
			case 4:
				exit = 1;
				break;
			// INVALID
			default:
				printf("Enter a valid option\n");
				break;
		}
	}
	
	return 0;
}

File: LinkedList/Circular LinkedList/Singly LinkedList/CSL.py
class Createnode:
    def __init__(self, data):
        self.data = data
        self.next = self


class CircularSinglyLinkedList:
    # Initializing head
    def __init__(self):
        self.head = None
        self.length = 0

    # Inserting new node
    def insert(self, index , data):
        # If list is empty
        if self.length==0 and index==1:
            self.head = Createnode(data)
            self.length += 1
            print(f"{data} is inserted at begining of list.")
        # if list is not empty
        else:
            # For valid index
            if index>0 and index<=self.length+1:

                newNode = Createnode(data)  # Create node
                temp = self.head
                
                if index == 1:      # For inserting at position 1
                    newNode.next = self.head
                    self.head = newNode
                    for i in range(1, self.length):	# Traverse to last node
                        temp = temp.next
                    temp.next = self.head
                else:               # For remaining position
                    for i in range(1, index-1):    # Traversing to index-1 node
                        temp = temp.next
                    newNode.next = temp.next
                    temp.next = newNode
                
                self.length += 1
                print(f"{data} is inserted at index {index}.")
                
            else:
                print("Invalid Index.")

    # Delete node
    def delete(self, index):
        if self.length == 0:    # if list is empty
            print("Empty list.")
        else:
            if index>0 and index<=self.length:

                if self.length==1 and index==1:	    # If length of list is one
                    temp = self.head
                elif index == 1:                    # For deleting at position 1
                    temp1 = self.head
                    for i in range(1, self.length):	    # Traverse to last node
                        temp1 = temp1.next
                    temp = temp1.next
                    temp1.next = temp.next
                    self.head = temp.next
                else:                               # For remaining position
                    temp1 = self.head
                    for i in range(1, index-1):         # Traversing to index-1 node
                        temp1 = temp1.next
                    temp = temp1.next
                    temp1.next = temp.next

                self.length -= 1
                print(f"{temp.data} is deleted from list.")
                
            else:
                print("Invalid Index.")

    # Traverse through list
    def display(self):
        if self.length == 0:    # if list is empty
            print("Empty list.")
        else:
            temp = self.head
            for i in range(1, self.length+1):
                print(temp.data, end="   ")
                temp = temp.next



menu = '''
Enter
    1. Insert
    2. Delete
    3. Display
    4. Exit
-> '''
if __name__ == "__main__":
    obj = CircularSinglyLinkedList()
    exit = False
    while not exit:
        switch_var = int(input(menu))
        # INSERT
        if switch_var == 1:
            temp = int(input("Enter the number : "))
            index = int(input(f"Enter the index(1-{obj.length+1}) : "))
            obj.insert(index, temp)
        # DELETE
        elif switch_var == 2:
            index = int(input(f"Enter the index(1-{obj.length}) : "))
            obj.delete(index)
        # DISPLAY
        elif switch_var == 3:
            obj.display()
        # EXIT
        elif switch_var == 4:
            exit = True
        # INVALID
        else:
            print("Please enter a valid statement.")